### ISSUE TITLE
5 point algorithm to compute Essential Matrix

### ISSUE THREAD

**Issue by ana42742** [scikit-image/scikit-image#6820]
https://github.com/scikit-image/scikit-image/blob/34ffe6a19611f07fac0e5ab6f912b09bb272c667/skimage/transform/_geometric.py#L501

In the `skimage.transform.EssentialMatrixTransform` class, we use the 8-point algorithm to estimate the essential matrix. There is a more efficient 5-point algorithm that could instead be used for the computation of the essential matrix. 

It has been seen that the 5-point algorithm estimate for the essential matrix usually has higher accuracy than the one that uses 8 point pairs. See [this paper](http://www.bmva.org/bmvc/2013/Papers/paper0116/paper0116.pdf). 

I did find some working [code](https://github.com/AugusteLef/Gravity_and_Relative_Pose_Estimation/blob/main/RelativePoseEstimation/5p_poselib.py) that uses the 5-point algorithm. 
Would it be possible/feasible to optimise this function?

**Comment by lagru** [comment]
Thanks for the suggestion @ana42742! I'll try to look into this tomorrow. Otherwise don't be afraid to ping me if I forget. :sweat_smile:

**Comment by ana42742** [comment]
Sure! @lagru :)

**Comment by rum1887** [comment]
> https://github.com/scikit-image/scikit-image/blob/34ffe6a19611f07fac0e5ab6f912b09bb272c667/skimage/transform/_geometric.py#L501
> 
> In the `skimage.transform.EssentialMatrixTransform` class, we use the 8-point algorithm to estimate the essential matrix. There is a more efficient 5-point algorithm that could instead be used for the computation of the essential matrix.
> 
> It has been seen that the 5-point algorithm estimate for the essential matrix usually has higher accuracy than the one that uses 8 point pairs. See [this paper](http://www.bmva.org/bmvc/2013/Papers/paper0116/paper0116.pdf).
> 
> I did find some working [code](https://github.com/AugusteLef/Gravity_and_Relative_Pose_Estimation/blob/main/RelativePoseEstimation/5p_poselib.py) that uses the 5-point algorithm. Would it be possible/feasible to optimise this function?

nice suggestion  @ana42742 , will have a look and get back to you  :)

**Comment by lagru** [comment]
Hey , I'm no expert in the topic and my background is not in image processing, but...  on the surface this looks like something worth investigating. From what I gather the paper has at least [some citations (13)](https://scholar.google.de/scholar?cites=13289613226626670225&as_sdt=2005&sciodt=0,5&hl=en) though from a few searches it seems like [Nistér 2004](https://ieeexplore.ieee.org/abstract/document/1288525) is way more established, e.g. also cited by [MATLAB](https://www.mathworks.com/help/vision/ref/estimateessentialmatrix.html#References). We usually aim to include well established algorithms. Though that's more of a subjective aim rather than a hard rule. 

Concerning the code you found, I can not find a license for it so right now we cannot include it. Though, we can always ask the author. And to be honest, I'm not sure if the code implements the 5pt algorithm from the paper; I can't find a reference to it in the repository.

@ana42742, I'm impressed and curious how you came about this idea, the paper and the code?

**Comment by ana42742** [comment]
@lagru Sorry for the delayed reply on this. Actually, while I was trying to add docstring examples as mentioned in #6761 to the EssentialMatrixTransform class, I looked into what essential and fundamental matrices were. I came across the 5-point algorithm through that research. Let me go through [Nistér 2004](https://ieeexplore.ieee.org/abstract/document/1288525) and get back to you. If it's implementable then we can go ahead.

**Comment by ana42742** [comment]
@lagru I have a couple of updates.
I found the implementation to estimate the Essential Matrix in MATLAB [see here](https://in.mathworks.com/matlabcentral/fileexchange/67580-essential-matrix-estimation). Upon downloading the files, I see that this implementation exists entirely in MATLAB with a heavy reliance on a C file (namely, `calibrated_fivepoint_helper.c`). I was wondering if there are any implementations in scikit-image that use C/C++ under the hood? Is it even possible for a skimage function to *call* a C program?

**Comment by lagru** [comment]
>  Is it even possible for a skimage function to call a C program?

Yes, we use a lot of C/C++ or Cython-based code in scikit-image (e.g. have a look for files ending with `.c` or `pyx`). 

We also have to make sure that the license is compatible to ours if we want to use it. I'm not sure but it looks like the MATLAB code covered by the BSD-3-clause license, which should be compatible.

**Comment by stefanv** [comment]
In this case, implementing with NumPy is much preferred over wrapping a C library.

You can already see a related algorithm implemented for the fundamental matrix.
